### PR TITLE
vdk-confluence: add data source plugin for confluence

### PR DESCRIPTION
**Pull request by duyguHsnHsn**
Objective: Develop a plugin for data sourcing from Confluence, incorporating an initial stream that comprehensively retrieves data across all Confluence spaces.

Linked to:  https://github.com/vmware/versatile-data-kit/issues/3048 (additional tickets are added here)

Signed-off-by: Duygu Hasan [hduygu@vmware.com](mailto:hduygu@vmware.com)

### CODE REVIEW

File: projects/vdk-plugins/vdk-confluence-data-source/tests/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Copyright 2021-2024 VMware, Inc.
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/tests/__init__.py:1]: remove the __init__.py file likelly it was generated by accident?

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/plugin_entry.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2021-2024 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+from typing import List
+
+from vdk.api.plugin.hook_markers import hookimpl
+from vdk.api.plugin.plugin_registry import IPluginRegistry
+from vdk.internal.core.config import ConfigurationBuilder
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/plugin_entry.py:7]: Fix codacy warnings particularly about unused imports

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+from vdk.plugin.data_sources.config import config_field
+from vdk.plugin.data_sources.data_source import DataSourcePayload
+from vdk.plugin.data_sources.data_source import IDataSource
+from vdk.plugin.data_sources.data_source import IDataSourceConfiguration
+from vdk.plugin.data_sources.data_source import IDataSourceStream
+from vdk.plugin.data_sources.factory import data_source
+
+
+@config_class(
+    name="confluence-data-source-config", description="Stream Confluence page contents"
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py:20]: I am pretty sure this should be the same as in data source name below.

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        default=True,
+    )
+    confluence_kwargs: Optional[dict] = config_field(
+        description="Additional keyword arguments " "for the Confluence client",
+        default=None,
+    )
+
+
+@data_source(
+    name="confluence-data-source", config_class=ConfluenceContentDataSourceConfiguration
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py:52]: The name should really be just "confluence" , data source is redundant in teh same


if you do `vdk data-sources --list`
```
git-data-source
confluence-data-source
```
looks a bit redundant and strange . You would expect just to see 
```
git
confluence
```

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                                        confluence_client=self._confluence_client,
+                                        space_key=space_key,
+                                    )
+                                )
+                except Exception as e:
+                    logging.error(f"Failed to fetch spaces from Confluence: {e}")
+                    raise
+
+    def disconnect(self):
+        self._streams = []
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py:157]: Do we need to close _confluence_client or something like that ?

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        "in dictionary format. Required keys: 'access_token',"
+        " 'access_token_secret', 'consumer_key', 'key_cert'",
+        default=None,
+    )
+    cloud: bool = config_field(
+        description="Flag indicating if the Confluence instance is" " cloud-based",
+        default=True,
+    )
+    confluence_kwargs: Optional[dict] = config_field(
+        description="Additional keyword arguments " "for the Confluence client",
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py:46]: what are those keyword arguments? What is this "confluence client" 

As a user reading the description I'd wonder those.

(hint: link to relevant sites may be ok)

File: projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        default=None,
+    )
+    oauth2: Optional[dict] = config_field(
+        description="OAuth2 credentials for authentication "
+        "in dictionary format. Required keys: 'access_token',"
+        " 'access_token_secret', 'consumer_key', 'key_cert'",
+        default=None,
+    )
+    cloud: bool = config_field(
+        description="Flag indicating if the Confluence instance is" " cloud-based",
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-confluence-data-source/src/vdk/plugin/confluence_data_source/data_source.py:42]: How do I know if my instance is cloud based

**Review by antoniivanov — APPROVED** [review]

File: projects/vdk-plugins/vdk-confluence-data-source/tests/script_for_manual_checks.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        for page_content in stream.read():
+            print(page_content.metadata)
+            count += 1
+            if count == 3:
+                break
+
+    confluence_data_source.disconnect()
+
+
+# run()
```
Comment (DeltaMichael) [inline projects/vdk-plugins/vdk-confluence-data-source/tests/script_for_manual_checks.py:43]: Nit: remove this.

**Review by DeltaMichael — APPROVED** [review]

**Pull request merged**